src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala

```diff
@@ -1,46 +1,70 @@
 
 
 trait StandardUserInfoDirectives extends UserInfoDirectives {
   implicit val executionContext: ExecutionContext
 
+  def requireUserInfo: Directive1[UserInfo] = (
+    headerValueByName(accessTokenHeader) &
+      headerValueByName(googleSubjectIdHeader) &
+      headerValueByName(expiresInHeader) &
+      headerValueByName(emailHeader)
+    ) tflatMap {
+    case (token, googleSubjectId, expiresIn, email) =>
+      onSuccess(Try(expiresIn.toLong).fold[Future[UserInfo]](t => Future.failed(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"expiresIn $expiresIn can't be converted to Long"))),
+        l => createUserInfo(OAuth2BearerToken(token), GoogleSubjectId(googleSubjectId), WorkbenchEmail(email), l, directoryDAO)))
+  }
+
+  def requireCreateUser: Directive1[CreateWorkbenchUserAPI] = (
+      headerValueByName(googleSubjectIdHeader) &
+      headerValueByName(emailHeader)
+    ) tflatMap {
+    case (googleSubjectId, email) => {
+      onSuccess(createWorkbenchUserAPI(GoogleSubjectId(googleSubjectId), WorkbenchEmail(email)))
+    }
+  }
+}
+
+object StandardUserInfoDirectives{
   val petSAdomain = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com".r
 
-  private def isPetSA(email: WorkbenchEmail) = {
+  def isPetSA(email: WorkbenchEmail) = {
     petSAdomain.pattern.matcher(email.value).matches
   }
 
-  def requireUserInfo: Directive1[UserInfo] = (
-    headerValueByName("OIDC_access_token") &
-      headerValueByName("OIDC_CLAIM_user_id") &
-      headerValueByName("OIDC_CLAIM_expires_in") &
-      headerValueByName("OIDC_CLAIM_email")
-    ) tflatMap {
-    case (token, userId, expiresIn, email) => {
-      val userInfo = UserInfo(OAuth2BearerToken(token), WorkbenchUserId(userId), WorkbenchEmail(email), expiresIn.toLong)
-      onSuccess(getUserFromPetServiceAccount(userInfo).map {
-        case Some(petOwnerUser) => UserInfo(OAuth2BearerToken(token), petOwnerUser.id, petOwnerUser.email, expiresIn.toLong)
-        case None => userInfo
-      })
-    }
+  def createUserInfo(token: OAuth2BearerToken, googleSubjectId: GoogleSubjectId, email: WorkbenchEmail, expiresIn: Long, directoryDAO: DirectoryDAO)(implicit ec: ExecutionContext): Future[UserInfo] = if (isPetSA(email)) {
+    // If it's a PET account, we treat it as its owner
+    directoryDAO.getUserFromPetServiceAccount(ServiceAccountSubjectId(googleSubjectId.value)).flatMap(_.fold[Future[UserInfo]](Future.failed(new WorkbenchException("no owner found for this PET account")))(pet => Future.successful(UserInfo(token, pet.id, pet.email, expiresIn.toLong))))
+  } else {
+    for{
+      subject <- directoryDAO.loadSubjectFromGoogleSubjectId(googleSubjectId)
+      userInfo <- subject match{
+        case Some(uid: WorkbenchUserId) => Future.successful(UserInfo(token, uid, email, expiresIn))
+        case Some(_) => Future.failed(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"subjectId $googleSubjectId is not a WorkbenchUser")))
```

Comment:
I was mostly following the other functions signatures such as https://github.com/broadinstitute/sam/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/DirectoryDAO.scala#L35...It also makes sense to me that a given googleSubjectId might be mapped to a non regular user subject. The case you pointed out is where we received a `googleSubjectId` for non regular user (say `WorkbenchGroup` ). I can change the return code to something else if you have a better one in mind.

Reply:
but the ldap schema does not allow `WorkbenchGroup` to have `googleSubjectId`.  `loadSubjectFromGoogleSubjectId` should never return anything other than a regular user so that function should be responsible for returning the error (and it is probably a 500). That relieves all the callers from having to do this.

Reply:
Okay, I can exclude `WorkbenchGroup` as an improvement....But it won't help this code here right, cuz it can still return a PET